### PR TITLE
Add a pax-exam based OSGi test suite

### PR DESCRIPTION
**Pull request by rovarga**
Problems with artifacts not being well-formed bundles havee been
reported, so create a unit test suite to catch the most obvious ones.

Signed-off-by: Robert Varga <nite@hq.sk>

### CODE REVIEW

**Comment by netkins** [comment]
**AUTOMATED MESSAGE for ADMINS**:<br>Please verify and accept the pull request.<br>To accept, say `@netkins accept`<br>To build again, say `@netkins build`<br>To whitelist the author, say `@netkins whitelist`

**Comment by normanmaurer** [comment]
@netkins accept

File: osgitests/pom.xml [hunk 1]
```diff
@@ -0,0 +1,221 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+  ~ Copyright 2012 The Netty Project
```
Comment (normanmaurer) [inline osgitests/pom.xml]: This should be 2015.

File: osgitests/src/test/java/io/netty/osgitests/OsgiBundleTest.java [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+import org.ops4j.pax.exam.Configuration;
+import org.ops4j.pax.exam.Option;
+import org.ops4j.pax.exam.junit.PaxExam;
+import org.ops4j.pax.exam.spi.reactors.ExamReactorStrategy;
+import org.ops4j.pax.exam.spi.reactors.PerMethod;
+
+@RunWith(PaxExam.class)
+public class OsgiBundleTest {
+    private static final String GROUP = "io.netty";
+    private static final Collection<String> BUNDLES = Arrays.asList(
```
Comment (normanmaurer) [inline osgitests/src/test/java/io/netty/osgitests/OsgiBundleTest.java]: I wonder if we can't read this out of the dependencies that are specified in the pom.xml. This would ensure we not miss to add something when adding a new module. WDYT ?

**Comment by normanmaurer** [comment]
@rovarga thanks for the work... please see my comments

File: osgitests/src/test/java/io/netty/osgitests/OsgiBundleTest.java (hunk 2 above)
Comment (rovarga) [inline osgitests/src/test/java/io/netty/osgitests/OsgiBundleTest.java]: Ah, maven/surefire/java interaction creativity ... target/classes/META-INF/maven/dependencies.properties should give everything necessary by the time unit tests run, so processing it should do the trick.

File: osgitests/pom.xml (hunk 1 above)
Comment (rovarga) [inline osgitests/pom.xml]: doh, corrected

**Comment by normanmaurer** [comment]
@rovarga could you also please ensure your commit message match our template:

http://netty.io/wiki/writing-a-commit-message.html

And please sign the ical:
http://tinyurl.com/netty-icla

**Comment by rovarga** [comment]
@normanmaurer  should be all set. let me know if there's anything else that needs addressing.

**Comment by normanmaurer** [comment]
@rovarga could you check why it failed: http://clinker.netty.io/jenkins/job/netty-pullreqs/601/io.netty$netty-osgitests/ ?

**Comment by rovarga** [comment]
@normanmaurer unit tests picked themselves as a dependency to check, and fails since this executes in the test phase, when the jar is non-existent yet. Fixed by waiting for the dependencies line.

**Comment by trustin** [comment]
I wonder if there's a way to test correctly even if a user specified the test phase only.

**Comment by rovarga** [comment]
@trustin not sure I follow, can you rephrase that? :)

**Comment by trustin** [comment]
Both `mvn test` and `mvn package` should test OSGi compatibility and the tests should pass. Did you already fix the failure? I don't see a follow-up commit after the build failure from CI.

**Comment by normanmaurer** [comment]
@trustin it was fixed... the test-failure is not related with this PR. I already cherry-picked to master, but not 4.1 and 4.0 so far.

**Comment by trustin** [comment]
Oh, I see. Thanks for clarification. Sorry about the noise, @rovarga. When would it be ported back to 4.1 and 4.0?

**Comment by trustin** [comment]
Oh, by the way, I was hoping to rename the module to `testsuite-osgi` or something like. Could you take care, @normanmaurer?

**Comment by normanmaurer** [comment]
@trustin sure can do...

**Comment by normanmaurer** [comment]
Cherry-picked into 4.0 (2aad79fcca5dee86ea2defb3daa66ee2d37196b6), 4.1 (fa646255221c6507d277ecd7630cf335d2e6045a) and master (d83700bd0a22dfb210dccb35eef4c6c7bcd19cdf and 6f03c15a93d0314c59dd0059ad1303f5ef3a0d8f)
